system/security/shared/caching.cpp

```diff
@@ -679,11 +681,97 @@ bool CPermissionsCache::queryPermsManagedFileScope(ISecUser& sec_user, const cha
     return rc;
 }
 
-SecAccessFlags CPermissionsCache::queryDefaultPermission(ISecUser& user)
+
+void CPermissionsCache::managedFileScopesCacheFillThread()
 {
-    if (!m_secMgr)
-        return SecAccess_Full; // if no security manager, all full access to all scopes
+    DBGLOG("CACHE: CPermissionsCache managedFileScopesCacheFillThread started");
+
+    // When the thread first starts, do a cache fill with the write lock held.
+    // This ensures any managed scope queries are locked out until the cache is filled.
+    {
+        WriteLockBlock writeLock(m_scopesRWLock);
+        m_fileScopeCacheFillSem.signal();  // we have started and have the lock to ensure the cache is filled
+        IArrayOf<ISecResource> scopes;
+        aindex_t count = m_secMgr->getManagedScopeTree(RT_FILE_SCOPE, nullptr, scopes);
+        removeAllManagedFileScopes();
+        if (count)
+            addManagedFileScopes(scopes);
+        time_t now;
+        time(&now);
+        m_lastCacheFillTime = now;
+    }
 
+    unsigned waitTimeSeconds = m_cacheTimeoutInSeconds;
+    while (!m_stopFileScopeCacheFillThread && m_cacheTimeoutInSeconds)
+    {
+        m_fileScopeCacheFillSem.wait(waitTimeSeconds * 1000);
```

Comment:
if this semaphore is signalled then you should break out of the loop immediately - you are terminating.

Reply:
Although the check on wait would be clearer, the subsequent check of m_stopFileScopeCacheFillThread means this thread will terminate correctly.

Reply:
Agreed. It won't cause problems, but it would be clearer to explicitly exit when signalled.